frame/tx-pause/src/lib.rs

```diff
@@ -267,21 +267,20 @@ impl<T: Config> Pallet<T> {
 	/// Return whether this pallet or call is paused
 	pub fn is_paused(full_name: &FullNameOf<T>) -> bool {
 		let (pallet_name, maybe_call_name) = full_name;
+		// SAFETY: Everything that is whitelisted cannot be paused,
+		// including calls within paused pallets.
+		if T::WhitelistCallNames::contains(&<FullNameOf<T>>::from((
+			pallet_name.clone(),
+			maybe_call_name.clone(),
+		))) {
+			return false
+		};
+		// Check is pallet is paused.
 		if <PausedCalls<T>>::contains_key(<FullNameOf<T>>::from((pallet_name.clone(), None))) {
-			// SAFETY: Everything that is whitelisted cannot be paused,
-			// including calls within paused pallets.
-			if T::WhitelistCallNames::contains(&<FullNameOf<T>>::from((
-				pallet_name.clone(),
-				maybe_call_name.clone(),
-			))) {
-				return false
-			};
 			return true
 		};
-		<PausedCalls<T>>::contains_key(<FullNameOf<T>>::from((
-			pallet_name.clone(),
-			maybe_call_name.clone(),
-		)))
+		// Check if call in a pallet is paused
+		<PausedCalls<T>>::contains_key(full_name)
```

Comment:
This is why benchmarking now [reports 2 reads](https://github.com/paritytech/substrate/pull/12092/commits/b8a4fcb5cf7625f97f689ec7f65ed48b97e48a0d#diff-8bdcea11c3a9b2b8dbd69638317db1d449279e05c739cba6a45d1455952f1399L83-R84).
As we have separate items for the pallet being paused vs. a specific call being paused via `FullNameOf`. Ideally only be one storage read, Perhaps instead of a double map we could use 
```diff
- 		StorageMap<_, Blake2_128Concat, (PalletNameOf<T>, Option<CallNameOf<T>>), (), OptionQuery>;
+		StorageMap<_, Blake2_128Concat, PalletNameOf<T>, Option<BoundedVec<CallNameOf<T>, MaxPalletCalls>>, OptionQuery>;
```

and then refactor to manage a a vec of calls in storage, this would allow for the pallet to be disabled (`None` for the value ) _or_ calls within it disabled (A vec of call names). Whitelist would remain the same, and still need to be checked first in `is_paused`.

Reply:
It is mission critical to make `is_paused` as light weight as possible as I believe this is checked for _every dispatchable_ that comes into the runtime, as BaseCallFilter will include this if the TxPause pallet is added :thinking:

Reply:
It is done every extrinsic, but the value is only read from the storage into the overlay once per block and all following extrinsics will re-use the overlay.  
So it looks bad in benchmarking but should be fine. Although I like that second approach more.  
`StorageMap<_, Blake2_128Concat, PalletNameOf<T>, BoundedVec<CallNameOf<T>, MaxPalletCalls>, ValueQuery>;`
is enough since a Vector has the semantically correct default in this case (nothing being paused).  
You could also do an Enum:  
```rust
enum PalletPauseStatus {
    /// All calls of the pallet are paused.
    AllCalls,
    /// Specific calls are paused.
    SpecificCalls(BoundedVec<…>>),
}
```
But just an idea. I dont want to keep extending the development time here :sweat_smile:

Reply:
Out of scope for now.